File: Day10-19/11.py
def manhatten_hex_dist(orig, dest):
    dist_x = dest[0] - orig[0]
    dist_y = dest[1] - orig[1]

    if dist_x > 0 == dist_y > 0:
        return abs(dist_x + dist_y)
    else:
        return max(abs(dist_x), abs(dist_y))


def next_loc(orig, direc):
    if direc == 'n':
        return (orig[0], orig[1]+1)
    elif direc == 'ne':
        return (orig[0]+1, orig[1])
    elif direc == 'se':
        return (orig[0]+1, orig[1]-1)
    elif direc == 's':
        return (orig[0], orig[1]-1)
    elif direc == 'sw':
        return (orig[0]-1, orig[1])
    elif direc == 'nw':
        return (orig[0]-1, orig[1]+1)


def path_dist(path):
    loc = (0, 0)
    m = 0
    for direc in path:
        loc = next_loc(loc, direc)
        m = max(m, manhatten_hex_dist((0, 0), loc))
    return manhatten_hex_dist((0, 0), loc), m


if __name__ == '__main__':
    with open('11.txt') as f:
        directions = f.read().split(',')
    part1, part2 = path_dist(directions)
    print("Part 1: {}".format(part1))
    print("Part 2: {}".format(part2))

File: Day10-19/11.go
package main

import (
	"fmt"
	"io/ioutil"
	"strings"
)

func abs(x int) int {
	if x < 0 {
		return -x
	}
	return x
}

func nextLoc(x int, y int, direction string) (int, int) {
	switch direction {
	case "n":
		return x, y + 1
	case "ne":
		return x + 1, y
	case "se":
		return x + 1, y - 1
	case "s":
		return x, y - 1
	case "sw":
		return x - 1, y
	case "nw":
		return x - 1, y + 1
	}
	panic("Invalid direction")
}

func hexManhattenDist(x0 int, y0 int, x1 int, y1 int) int {
	distX := x1 - x0
	distY := y1 - y0

	if (distX > 0) == (distY > 0) {
		return abs(distX + distY)
	} else if abs(distX) > abs(distY) {
		return abs(distX)
	} else {
		return abs(distY)
	}
}

func calculatePathDistances(directions []string) (endDist int, maxDist int) {
	x := 0
	y := 0
	maxDist = 0
	for _, direction := range directions {
		x, y = nextLoc(x, y, direction)
		dist := hexManhattenDist(0, 0, x, y)
		if dist > maxDist {
			maxDist = dist
		}
	}
	endDist = hexManhattenDist(0, 0, x, y)
	return
}

func main() {
	dat, _ := ioutil.ReadFile("./11.txt")
	directions := strings.Split(string(dat), ",")
	part1, part2 := calculatePathDistances(directions)
	fmt.Printf("Part 1: %d\n", part1)
	fmt.Printf("Part 2: %d\n", part2)
}
